metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/TaxonomyEventHandler.java

```diff
@@ -0,0 +1,149 @@
+        String classTaxName = findTaxName(domQueryUsingXPath, CLASS);
+        Taxonomy classTaxonomy = null;
+        if (classTaxId != null && !classTaxId.equals("")) {
+            long classTaxIdLong = Long.parseLong(classTaxId);
+            classTaxonomy = taxonomyRepository.findByTaxonomyId(classTaxIdLong);
+            if (classTaxonomy == null) {
+                classTaxonomy = taxonomyRepository.save(new Taxonomy(classTaxIdLong, classTaxName, CLASS));
+            }
+            taxonomy.setTaxonomyClass(classTaxonomy);
+        }
+
+        String orderTaxId = findTaxId(domQueryUsingXPath, ORDER);
+        String orderTaxName = findTaxName(domQueryUsingXPath, ORDER);
+        Taxonomy orderTaxonomy = null;
+        if (orderTaxId != null && !orderTaxId.equals("")) {
+            long orderTaxIdLong = Long.parseLong(orderTaxId);
+            orderTaxonomy = taxonomyRepository.findByTaxonomyId(orderTaxIdLong);
+            if (orderTaxonomy == null) {
+                orderTaxonomy = new Taxonomy(orderTaxIdLong, orderTaxName, ORDER);
+                orderTaxonomy.setTaxonomyClass(classTaxonomy);
+                orderTaxonomy = taxonomyRepository.save(orderTaxonomy);
+            }
+            taxonomy.setTaxonomyOrder(orderTaxonomy);
+        }
+
+        String genusTaxId = findTaxId(domQueryUsingXPath, GENUS);
+        String genusTaxName = findTaxName(domQueryUsingXPath, GENUS);
+        Taxonomy genusTaxonomy = null;
+        if (genusTaxId != null && !genusTaxId.equals("")) {
+            long genusTaxIdLong = Long.parseLong(genusTaxId);
+            genusTaxonomy = taxonomyRepository.findByTaxonomyId(genusTaxIdLong);
+            if (genusTaxonomy == null) {
+                genusTaxonomy = new Taxonomy(genusTaxIdLong, genusTaxName, GENUS);
+                genusTaxonomy.setTaxonomyClass(classTaxonomy);
+                genusTaxonomy.setTaxonomyOrder(orderTaxonomy);
+                genusTaxonomy = taxonomyRepository.save(genusTaxonomy);
+            }
+            taxonomy.setTaxonomyGenus(genusTaxonomy);
+        }
+
+        String speciesTaxId = findTaxId(domQueryUsingXPath, SPECIES);
+        String speciesTaxName = findTaxName(domQueryUsingXPath, SPECIES);
+        Taxonomy speciesTaxonomy = null;
+        if (speciesTaxId != null && !speciesTaxId.equals("")) {
+            long speciesTaxIdLong = Long.parseLong(speciesTaxId);
+            speciesTaxonomy = taxonomyRepository.findByTaxonomyId(speciesTaxIdLong);
+            if (speciesTaxonomy == null) {
+                speciesTaxonomy = new Taxonomy(speciesTaxIdLong, speciesTaxName, SPECIES);
+                speciesTaxonomy.setTaxonomyClass(classTaxonomy);
+                speciesTaxonomy.setTaxonomyOrder(orderTaxonomy);
+                speciesTaxonomy.setTaxonomyGenus(genusTaxonomy);
+                speciesTaxonomy = taxonomyRepository.save(speciesTaxonomy);
+            }
+            taxonomy.setTaxonomySpecies(speciesTaxonomy);
+        }
```

Comment:
Currently this method is a bit lengthy and uses lots of copy-paste, triggering very similar operations for class, order, genus, and then species. I suggest rewriting it in a recursive fashion:

* Given the taxonomy, try to fetch existing by ID and return it
* Otherwise, for each of the fixed-rank methods, call: `taxonomy.setTaxonomyClass(new Taxonomy(classId))` (and similar for order, genus, and species).
  + This will again trigger this same function, which will deal with the rest of the process for that taxonomy level
  + If some fixed-ranks for this level do not exist, they should be populated by `null` (e. g. when importing an order, there will be no species ID in the lineage)
* Save the resulting entity to repository

This will simplify both the function code and the addition of new fixed ranks in the future

Reply:
ya tried doing it  i too thought the same but guess we needed to use reflection to differntiate setTaxonomyClass setTaxonomyGenus and setTaxonomyOrder method functions

Reply:
ya i undertand what you said , let me check if we can do it

Reply:
Cool, much better now! May I suggest one final modification: we can add `setTaxonomyForRank(Taxonomy taxonomy, String rank)` inside `Taxonomy.java` which would accept taxonomy entity and rank, and based on a rank would call the appropriate method (e. g. `setTaxonomyClass`, `setTaxonomyGenus` etc.) This would allow us to collapse this section into a single block, being executed for CLASS, ORDER, GENUS, and SPECIES, without using reflection.

Reply:
ya it better now thank :)

Reply:
Perfect now!